File: 프로그래머스/lv3/43162. 네트워크/네트워크.py
def solution(n, computers):
    
    def find(x):
        if parent[x] != x:
            parent[x] = find(parent[x])
        return parent[x]
    
    def union(a, b):
        a = find(a); b = find(b)
        if a < b: parent[b] = a
        else: parent[a] = b
        
    parent = [_ for _ in range(n)]
    for i in range(n):
        for j in range(n):
            if computers[i][j] == 1:
                if find(i) != find(j): union(i, j)
    
    for i in range(n): find(i) # parent 한 번 갱신 해줘야함
    
    return len(set(parent))

File: 프로그래머스/3/43162. 네트워크/네트워크.cpp
#include <string>
#include <vector>
#include <iostream>
#include <set>

using namespace std;

vector<int> parent;

int Find(int x)
{
    if (parent[x] != x)
    {
        parent[x] = Find(parent[x]);
    }
    return parent[x];
}

void Union(int a, int b)
{
    a = Find(a); b = Find(b);
    if (a < b)
    {
        parent[b] = a;
    }
    else
    {
        parent[a] = b;
    }
}

int solution(int N, vector<vector<int>> computers) {
    
    for (int i = 0; i < N; ++i)
    {
        parent.push_back(i);
    }
    
    for (int i = 0; i < N; ++i)
    {
        for (int j = i+1; j < N; ++j)
        {
            if (computers[i][j] == 1)
            {
                if (Find(i) != Find(j))
                {
                    Union(i, j);
                }
            }
        }
    }
    
    for (int i = 0; i < N; ++i)
    {
        Find(i);
    }
    
    
    set<int> parent_set(parent.begin(), parent.end());
    
    return parent_set.size();
}